transformer update

StireArLankar opened the pull request:
Hello sir, thank you for creating this library as it helped me with my work (linking existing db and tables in client), however i want to help you with handling recursive types.

### Description
1. As mentioned in `zod` description, best way to handle recusive types is to privide implicit typing.
And output of prisma generator is exactly what is needed, so i suggest to import them and type `zod`-validators with it as shown below
```ts
import { z } from 'zod';
import { NestedBoolFilterObjectSchema } from './NestedBoolFilter.schema';

import type { Prisma } from '@prisma/client';

const Schema: z.ZodType<Prisma.BoolFilter> = z
  .object({
    equals: z.boolean().optional(),
    not: z
      .union([z.boolean(), z.lazy(() => NestedBoolFilterObjectSchema)])
      .optional(),
  })
  .strict();

export const BoolFilterObjectSchema = Schema;
```
2. To fix any recursive imports problem wrap every ObjectSchema with `z.lazy`

----

Also there were some things i suggest to change:
1. Use strict on every `ObjectSchema`, because without it unions with objects, where all fields are optional, work incorrectly
```ts
const union = z.union([
    z.object({ foo: z.string().optional() }),
    z.object({ bar: z.string().optional() })
])

const res = union.parse({ bar: "" }) // {}, 
```
`bar` is stripped because it passes in first union member and default zod behaviour is to strip unknown fields
so with `.strict()` it will force zod to validate inputs until correct union member is found

2. Small refactoring in `Transformer` class - default properties values in constructor and types fixes overall 

----

I used generator with my improvements on db with 17 tables and haven't yet found any problems (tho i am not sure how lines like `field.includes('Filter')` we working before)

P.S.: maybe in README should be suggested to git-ignore all generated schemas, because in my case i got 1k+ files...

omar-dulaimi commented:
Hey @StireArLankar 

Thank you so much for your contribution. You've actually put a lot of thought and effort into this, and I really appreciate it.

Did you get a chance to test your work on a couple of schemas? like some Json fields and others.

Also, I added some feedback for you, make sure to check it out.

I have read all your comments in the code and I understood them well. Could you remove them?

Thanks again.

StireArLankar commented:
Hello @omar-dulaimi .

> 
> Did you get a chance to test your work on a couple of schemas? like some Json fields and others.
> 

I checked `Json` field and added with latest commit few lines of code so generated `JsonNullableFilter` and `JsonNullableWithAggregatesFilter` schemas now match prisma`s types.

I tested it for case with `previewFeatures = ["filterJson"]` enabled (somehow it says that it is deprecated for me, but generated prisma types miss `path` field and others) with mysql DB.

Something like this correctly filtered data, had autocompletion and passed thru zod validation
```ts
const where = {
    jsonField: {
        path: "$.boolean_inside_json",
        equals: false,
    },
}
```

For others fields i haven't really found any issues (except unions with optional objects stripping fields, but as i wrote in first comment i fixed it using`.strict()`).
Not like i heavily used it in my projects...

> 
> I have read all your comments in the code and I understood them well. Could you remove them?
> 

Removed all mine comments.

>
> Also, I added some feedback for you, make sure to check it out.
>

Not sure where i can find it?

omar-dulaimi commented:
Those are the comments I left for you as you can see in the screenshot, I'd love if you answered them.
![Screenshot from 2022-07-02 22-46-19](https://user-images.githubusercontent.com/11743389/177014419-2f7cd370-3720-46c2-8353-2d0c4c106211.png)

omar-dulaimi commented:
I added a field to the schema like this:

```
model Post {
   meta: Json
}
```

What was generated is this:

```
meta: z.lazy(() => JsonNullValueInputSchema)
```

So, I have two questions:

1- Do we really need to use `lazy` here?
2- According to Prisma Client, the `meta` field should have this type:
```ts
meta: JsonNullValueInput | InputJsonValue
```

The first type above is generated and has no problem, the second though wasn't. Here's how it looks:

- Found inside `node_modules/.prisma/client/index.d.ts`
```ts
type InputJsonValue = string | number | boolean | InputJsonObject | InputJsonArray
```

Do you have an idea on how to tackle this?

StireArLankar commented:
>
> Those are the comments I left for you as you can see in the screenshot, I'd love if you answered them.
>
Sorry, but i really can`t find those comments, i didn't get any notifications, probably it has something to do with `review` mechanism, maybe because you are the only reviewer i see nothing. 
Or im just blind. ¯\_(ツ)_/¯ 

Atleast i can answer those from screenshot here:
1. Array<string> and string[] are exact same thing, and as you had this in class props
```ts
  fields?: PrismaDMMF.SchemaArg[];
  ...
  modelOperations?: PrismaDMMF.ModelMapping[];
  enumTypes?: PrismaDMMF.SchemaEnum[];
```
 i thought you copied `Array<string>` from somewhere else and i simply replaced all occurrences
 
 2. Moved with last [commit](https://github.com/omar-dulaimi/prisma-zod-generator/pull/1/commits/acd48570f3ffeca6371de440cb2d22fd5a5cae58)
 3. Moved with last [commit](https://github.com/omar-dulaimi/prisma-zod-generator/pull/1/commits/acd48570f3ffeca6371de440cb2d22fd5a5cae58)
 
>
>So, I have two questions:
>
> 1- Do we really need to use lazy here?
> 2- According to Prisma Client, the meta field should have this type:
>

1. Not really, but generation of this part uses same code as generation with any other `Schema`
And problem with recursive types is not only because schema can reference itself, but because schema can reference other schema, which in other way can reference first one (this happens in all relation filters). So you get circular imports and in one schema field will be undefined (as i understand, this is the cause for problem that you tried to fix with `?.optional()`).

So it just easier to wrap any schema with `z.lazy()` and forget about trying to find where it should really be.

2. Did you try with this [commit](https://github.com/omar-dulaimi/prisma-zod-generator/pull/1/commits/373e1208179183c48498f45ca136cab4e26511ce) ?

Because i am getting for
```prisma
model Post {
  id        Int      @id @default(autoincrement())
  meta      Json
}
```
in `PostCreateInput.schema.ts` this
```ts
import { z } from 'zod';
import { JsonNullValueInputSchema } from '../enums/JsonNullValueInput.schema';

import type { Prisma } from '@prisma/client';

const literalSchema = z.union([z.string(), z.number(), z.boolean()]);
const jsonSchema: z.ZodType<Prisma.InputJsonValue> = z.lazy(() =>
  z.union([
    literalSchema,
    z.array(jsonSchema.nullable()),
    z.record(jsonSchema.nullable()),
  ]),
);

const Schema: z.ZodType<Prisma.PostCreateInput> = z
  .object({
    meta: z.union([z.lazy(() => JsonNullValueInputSchema), jsonSchema]),
  })
  .strict();

export const PostCreateInputObjectSchema = Schema;
```

omar-dulaimi commented:
> Not really, but generation of this part uses same code as generation with any other Schema
> And problem with recursive types is not only because schema can reference itself, but because schema can reference other schema, which in other way can reference first one (this happens in all relation filters). So you get circular imports and in one schema field will be undefined (as i understand, this is the cause for problem that you tried to fix with ?.optional()).
> So it is just easier to wrap any schema with z.lazy() and forget about trying to find where it should really be.

Awesome!



> Did you try with this [commit](https://github.com/omar-dulaimi/prisma-zod-generator/pull/1/commits/373e1208179183c48498f45ca136cab4e26511ce) ?

I did, but I still get this:

```ts
import { z } from 'zod';
import { JsonNullValueInputSchema } from '../enums/JsonNullValueInput.schema';

import type { Prisma } from '@prisma/client';

const Schema: z.ZodType<Prisma.PostCreateInput> = z
  .object({
    meta: z.lazy(() => JsonNullValueInputSchema),
  })
  .strict();

export const PostCreateInputObjectSchema = Schema;
```

omar-dulaimi commented:
> I used generator with my improvements on db with 17 tables and haven't yet found any problems (tho i am not sure how lines like field.includes('Filter') we working before)

It was working through the magic of workarounds. How you do you test the generated schemas? Do you have an API of some sort?


> P.S.: maybe in README should be suggested to git-ignore all generated schemas, because in my case i got 1k+ files...

I'll make sure to include that in the next release.

StireArLankar commented on src/transformer.ts:
```diff
@@ -3,30 +3,20 @@ import path from 'path';
-    this.modelOperations = modelOperations ?? [];
-    this.schemaImports = new Set();
-    this.enumTypes = enumTypes;
+  hasJson = false;
```
made it private with latest [commit](https://github.com/omar-dulaimi/prisma-zod-generator/pull/1/commits/78d7a4fc1b47239d0edf9055762f4109cbc77fbc)

i set `hasJson` to true inside `getObjectSchemaLine` function when at least one field type is `Json`, and inside `getFinalForm` add if it's `true` these lines
```ts
const literalSchema = z.union([z.string(), z.number(), z.boolean()]);
const jsonSchema: z.ZodType<Prisma.InputJsonValue> = z.lazy(() =>
  z.union([
    literalSchema,
    z.array(jsonSchema.nullable()),
    z.record(jsonSchema.nullable()),
  ]),
);
```
so i think is works on schema level?

StireArLankar commented:
>
> I did, but I still get this:
>

I can't understand why, maybe it has something to do with prisma's versions? During `npm i` my npm completly overrides `package-lock.json` because it is higher version than yours (i didn't include it in commit), and also i had to install latest version of `@prisma/sdk` because locked one didn't have types somehow and ts during compilation was angry. 

Other than that i don't know what can differ between code in pull request and my local machine.

>
> It was working through the magic of workarounds. How you do you test the generated schemas? Do you have an API of some sort?
>

I made few pages with tables and checked basic features (like pagination, sorting, filtering by few fields), so it was pretty much shallow check. It's possible to write some tests, where shemas will be generated in `beforAll`, dynamicly imported and used on some predefined inputs and cleared `afterAll`, but it's really a lot of work.

And tbh, i had some thought on sunday and came to conclusion that for my work tasks i can cheat a bit - because chain of

- i want to have types on client
- i put zod validators in trpc handlers so client can infer types
- i generate zod schemas based on db models
- i explicitly type schemas because zod needs help with recursive types
- i use types, generated by prisma

can be shortened to

- i want to have types on client
- i use as input `z.object({}).passthrough`
- i use prisma generated types to tell trpc what input i want - `... as z.ZodType<Prisma....Input>` with some omitted fields

and leave validation to prisma client (as it also does it)

omar-dulaimi commented:
> I can't understand why, maybe it has something to do with prisma's versions? 
> Other than that i don't know what can differ between code in pull request and my local machine.

Checking now actually the issue is solved. I can see the generated Json type.

> During npm i my npm completly overrides package-lock.json because it is higher version than yours (i didn't include it in commit), and also i had to install latest version of @prisma/sdk because locked one didn't have types somehow and ts during compilation was angry.

Yeah, I'll update package-lock. I had an issue with nvm before and it was breaking node for me(like 18 and 17). But I have fixed it.



> I made few pages with tables and checked basic features (like pagination, sorting, filtering by few fields), so it was pretty much shallow check.

I tested createOne and findMany, they appear to be working fine. No typescript errors or warnings are visible to me.


> It's possible to write some tests, where shemas will be generated in beforAll, dynamicly imported and used on some predefined inputs and cleared afterAll, but it's really a lot of work.

That is something I'm very interested in implementing for all my generators. I have never done real testing before, so it's going to take me some time to get it right. I would appreciate any pointers you may have.



> can be shortened to
> 
> i want to have types on client
> i use as input z.object({}).passthrough
> i use prisma generated types to tell trpc what input i want - ... as z.ZodType<Prisma....Input> with some omitted fields

Do you have a sample repo/code for that? Seems interesting honestly.


Also, I think I'm ready to merge. Do you have any more notes or changes?

StireArLankar commented:
> Do you have a sample repo/code for that? Seems interesting honestly.

I made example in this [branch](https://github.com/StireArLankar/prisma-zod-generator/tree/test)
Both queries with input validation (`test1`) and without (`test2`) have autocompletions and both throw on invalid input

> Also, I think I'm ready to merge. Do you have any more notes or changes?

I don't think i have anything else

omar-dulaimi commented:
Thanks a lot @StireArLankar for your work on this PR.

I'll make sure to check your branch.
